### PR TITLE
feat(spec): Add API implementation

### PR DESCRIPTION
**Pull request by josecolella**
⚠️ Best reviewed commit-by-commit

## This PR

- This PR adds the API object that implements the specifications from [api-initialization-and-configuration](https://docs.openfeature.dev/docs/specification/sections/flag-evaluation#11-api-initialization-and-configuration)
- This PR *does not* implement anything else, as I would like to break up the implementations into several PR for clearer intent

### Related Issues

References #7 

### Notes

- Add concurrent-ruby for thread safe data structures

### Follow-up Tasks

- [ ] Add client implementation

### How to test

1. Pull down branch
2. Run `bundle exec rspec`

### CODE REVIEW

File: lib/openfeature/sdk/api.rb [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+module OpenFeature
+  module SDK
+    # API Initialization and Configuration
+    #
+    # Represents the entry point to the API, including configuration of <tt>Provider</tt>,<tt>Hook</tt>,
+    # and building the <tt>Client</tt>
+    #
+    # To use the SDK, you can optionally configure a <tt>Provider</tt>, with <tt>Hook</tt>
+    #
+    #   OpenFeature::SDK.configure do |config|
```
[the remark this answers is not in the document]
Comment (josecolella) [reply lib/openfeature/sdk/api.rb:20]: The documentation needs to be updated to reflect this new structure

[the remark this answers is not in the document]
Comment (toddbaert) [reply lib/openfeature/sdk/api.rb:20]: > Not sure though, why couldn't we make API a module?

My understanding is that classes are literally a subclass of modules in Ruby. But there's a lot of state to maintain on the API singleton, so class seems right to me.

I'm not a Ruby person though, so please take that into account.

[the remark this answers is not in the document]
Comment (josecolella) [reply lib/openfeature/sdk/api.rb]: @mschoenlaub Just fixed documentation. Let me know if there are any other concerns

File: Gemfile.lock [hunk 2]
```diff
@@ -55,6 +56,7 @@ PLATFORMS
   x86_64-linux
 
 DEPENDENCIES
+  concurrent-ruby
```
Comment (toddbaert) [inline Gemfile.lock:59]: This is one of the most popular ruby deps, but it's runtime, so if we can avoid it that's a plus. 

It seems like it provides concurrent primitives (arrays, etc), so it's probably worth it.

**Review by toddbaert — APPROVED** [review]
This looks good to me! Thanks for keeping it digestible.

File: Gemfile.lock (hunk 2 above)
Comment (josecolella) [reply Gemfile.lock:59]: yeah, I was looking into the thread safetyness, and Ruby does provide [Mutex](https://ruby-doc.org/core-3.1.2/Thread/Mutex.html). With `concurrent-ruby` you pretty much get all the data structures in a thread-safe wrapped manner

**Comment by josecolella** [comment]
Team going to merge this, but please let me know if there are any follow ups. I will continue to provide easy to review PR instead of everything in one PRs

**Pull request merged**